### ISSUE TITLE
Make things a lil more mouse friendly.

### ISSUE THREAD

**Issue by keshto** [riverwm/river#526]
I understand things are supposed to be keyboard driven most of the time. However, sometimes I have a hand on the mouse and would like to control things more with the mouse. Some ideas:
1. Moving windows to different positions in the layout would be great.
2. Trying to drag a floating window from one monitor(output) to another does not work :(
3. if (1) is implemented be sure to allow dragging windows to other output's layouts.
4. resize windows with a border drag (no mod keys)
5. would be nice to have some sort of (thicker?) border actions, so one could for example, middle click close a window.
6. since the windows have gaps, clicking on the root to invoke a program or do an action would be nice for quick popup menus (launchers) etc.

**Comment by Leon-Plickat** [comment]
> Moving windows to different positions in the layout would be great.

How would river differentiate between dragging a window to change its position in the layout and dragging a window to make it floating?

> Trying to drag a floating window from one monitor(output) to another does not work :(

Windows being bound to outputs is (mostly) intentional. This may change in the future.

> resize windows with a border drag (no mod keys)

Depends on #399.

> would be nice to have some sort of (thicker?) border actions, so one could for example, middle click close a window.

Please open a separate issue for this, bulk issues like this one are rather impractical for tracking feature requests and progress.

> since the windows have gaps, clicking on the root to invoke a program or do an action would be nice for quick popup menus (launchers) etc.

This can already be implemented with a special client.

**Comment by keshto** [comment]
will open new issues for each that need haven't been addressed in your last reply.

**Comment by keshto** [comment]
> > Trying to drag a floating window from one monitor(output) to another does not work :(
> 
> Windows being bound to outputs is (mostly) intentional. This may change in the future, depending on how soon river adopts wlroots scene graph API.
Curious on why that was intentional? It is very un-intuitive imho.

> > since the windows have gaps, clicking on the root to invoke a program or do an action would be nice for quick popup menus (launchers) etc.
> 
> This can already be implemented with a special client.
Special client? like rivertile or?

**Comment by Leon-Plickat** [comment]
> Curious on why that was intentional? It is very un-intuitive imho.

There are many reasons.

I implemented moving and resizing windows via mouse before output layouts were configurable, so moving views between outputs via mouse simply wasn't a concern yet. Also it is a special case that requires rendering a window on two outputs at once and I wanted to avoid that complexity. Since moving a window out of an output and it not appearing on the adjacent output would probably confusing and since I already think moving windows outside the constraints of an output is an anti-pattern that should be avoided, I decided to just not allow it.

> Special client? like rivertile or?

A client that uses the layershell to provide a surface on the "desktop" so it can react when you click on it.

**Comment by keshto** [comment]
> > Curious on why that was intentional? It is very un-intuitive imho.
> 
> There are many reasons.
> 
> I implemented moving and resizing windows via mouse before output layouts were configurable, so moving views between outputs via mouse simply wasn't a concern yet. Also it is a special case that requires rendering a window on two outputs at once and I wanted to avoid that complexity. Since moving a window out of an output and it not appearing on the adjacent output would probably be confusing and since I already think moving windows outside the constraints of an output is an anti-pattern that should be avoided, I decided to just not allow it.
> 
> Since different outputs already are somewhat disconnected by having a separate set of tags, it mostly fits. There are different ways of making something "intuitive". River being consistent with itself is important, river being consistent with other desktops is not.

Hmm. Makes sense why it was intentional from a technical perspective. From a UX not so much so. A way to solve this is that currently a window does not go out of bounds on the output (good, keep that. Although I do wonder what happens when the window is bigger than the output and windows are not movable/rendered off screen). The mouse cursor when dragging a window currently gets locked into place when the window reaches the end of the screen. This should not be the case and the cursor should allow to continue to move and once the cursor moves to another output then send the window that is being moved to the other output. That way one does not need to implement rendering a window on 2 different outputs and still gets to drag a window as one would expect.

> A client that uses the layershell to provide a surface on the "desktop" so it can react when you click on it.

Ah, the layershell. Love how river tries to just do what it is supposed to do :+1: 


Sidenote: might want to give a few days before cowboy closing issues just in case there is more discussion...

**Comment by Leon-Plickat** [comment]
> The mouse cursor when dragging a window currently gets locked into place when the window reaches the end of the screen. This should not be the case and the cursor should allow to continue to move [...]

FWIW this is also intentional. It has always been a pet-peeve of mine that when you move things with the mouse inside constraints, that only that object respects the constraints and the cursor happily continues to wiggle around. That always felt wrong to me. So I decided to fix that while I was already getting my hands dirty in cursor code and just attach the cursor the object - a window in this case - statically. From my POV this is a substantial UX upgrade.

**Comment by Leon-Plickat** [comment]
> [...] as one would expect.

As far as I am concerned, this is just a synonym for "we have always done it like this", which I do not accept as a reasonable argument. I do not really care if river breaks with what other desktops have done, otherwise river would be a windows 98 clone.

If you have good arguments for this feature, I would definitely be interested in reading them, but as far as expectations go, I consider only internal consistency relevant.
